docs/edge_migration.md

```diff
@@ -0,0 +1,73 @@
+```
+
+We continue to write the Hindsight PubSub topic from Phase 1, but we move the HTTP output AWS-side. This will help us better empirically determine performance and various other implications of cross-cloud requests without potentially affecting production AWS processing. We can still use the GCP pubsub topic for stage testing, but it won't necessarily be actively used and the production GCP Edge will be receiving its data directly from AWS via HTTP POST.
+
+## Phase 3
+
+Timeline: **2019**
+
+```
+Data Producers -> AWS Tee -> AWS Edge
+                          \
+                           `-> GCP Edge
+```
+
+This is how we did the last major migration from Heka to Hindsight.
+
+This architecture introduces risk of data loss, so should be considered more dangerous than previous phases. This is why we are not planning on doing this until we're reasonably confident in the efficacy of the GCP infrastructure. In active tee mode, the client will be affected by GCP edge processing, particularly in request processing time and potentially by its status code. Depending on how we configure the tee, the AWS data ingestion infrastructure is susceptible to data duplication or loss due to client retry behavior. 
+
+We should ensure that we are sufficiently confident in the behaviour, performance, and stability of the GCP Edge before we move to this phase to ensure things don't go south. The previous phases are safer and should let us discover any major issues before we proceed to this phase.
+
+This does the "last mile" testing of most of the major missing pieces from earlier phases, in addition to prepping for the eventual fourth phase.
+
+## Phase 3 (alternative)
+
+```
+Data Producers -> Weighted DNS -> AWS Edge
+                               \
+                                `-> GCP edge
+```
+
+This alternative does not make use of a Tee to duplicate traffic.
+
+Depending on the results of Phase 2, an alternative strategy to teeing would be weighted DNS and running two GCP edges. In this configuration we could e.g. tee 1% of data directly to one of the GCP edges, while having the other 99% be processed by the other edge as per earlier phases. We would then need to do the reverse of Phase 1 and write that 1% back to AWS + Kafka for the AWS ingestion stack to process. This strategy can be just as dangerous as using the openresty tee because data loss on either side may result in partial data to both, and also requires writing code to convert the GCP representation back to the AWS representation. The risk can be mitigated by the DNS weights i.e. the amount being sent directly to GCP, which is an advantage. This is an interesting variation more similar to a standard DNS cutover, if required.
+
+## Phase 4
+
+Timeline: **TBD**
+
+```
+Data Producers -> GCP Edge
+```
+
+This will happen when we are confident that there is no risk of data loss by switching the endpoint to the GCP stack.
```

Comment:
I'm not sure this is an accurate characterization of why Phase 4 isn't going to happen for a while. I think we will be reasonably certain that we won't be losing data on the GCP infra long before we've fully migrated off all the AWS-based infrastructure and data processing, but I might be wrong.

Reply:
My intention here was "this won't happen *until* we are confident bla bla". It may be significantly later, depending on the realities of sunsetting things in AWS. I'll clarify.